stl/inc/bit

```diff
@@ -94,20 +95,128 @@ _NODISCARD constexpr _Ty rotr(const _Ty _Val, const int _Rotation) noexcept {
     }
 }
 
-template <class _Ty, enable_if_t<_Is_standard_unsigned_integer<_Ty>, int> _Enabled>
-_NODISCARD constexpr int countl_zero(const _Ty _Val) noexcept {
+
+inline bool __Is_popcnt_supported() {
+    int cpuinfo[4];
+    __cpuid(cpuinfo, 0x01);
+    return cpuinfo[2] & (1 << 23);
+}
+
+inline bool _Is_lzcnt_supported() {
+    int cpuinfo[4];
+    __cpuid(cpuinfo, 0x80000001);
+    return cpuinfo[2] & (1 << 5);
+}
+
+#pragma warning(push)
+#pragma warning(disable : 4640)
+
+template <typename _Ty>
+constexpr int _Popcount_helper(_Ty _Val) {
     constexpr int _Digits = numeric_limits<_Ty>::digits;
-    if (_Val == 0) {
+    // we static cast these bit patterns in order to truncate
+    // them to the correct size
+    _Val = static_cast<_Ty>(_Val - ((_Val >> 1) & static_cast<_Ty>(0x5555555555555555ull)));
+    _Val = static_cast<_Ty>(
+        (_Val & static_cast<_Ty>(0x3333333333333333ull)) + ((_Val >> 2) & static_cast<_Ty>(0x3333333333333333ull)));
+    _Val = static_cast<_Ty>((_Val + (_Val >> 4)) & static_cast<_Ty>(0x0F0F0F0F0F0F0F0Full));
+    for (int _ShiftDigits = 8; _ShiftDigits < _Digits; _ShiftDigits <<= 1) {
+        _Val = static_cast<_Ty>(_Val + static_cast<_Ty>(_Val >> _ShiftDigits));
+    }
+    // we want the bottom "slot" that's big enough to store _Digits
+    return static_cast<int>(_Val & static_cast<_Ty>(_Digits + _Digits - 1));
+}
+
+#if defined(_M_IX86) || defined(_M_X64)
+
+template <typename _Ty>
+inline int _Checked_x86_x86_64_countl_zero(_Ty _Val) {
+    static bool _Have_lzcnt = _Is_lzcnt_supported();
+    constexpr int _Digits   = numeric_limits<_Ty>::digits;
+    // lzcnt (when it doesn't fall back to bsr) is defined correctly for zero
+    if (!_Have_lzcnt && _Val == 0) {

```

Comment:
Is there a benefit to this check?

I could imagine that we could possibly use some `if constexpr` here, but right now this just complicates the code for no benefit as the result will always be the same

Reply:
Do you mean to replace `if (!_Have_lzcnt && _Val == 0)` with `if (_Val == 0)` ?
It is worse for branch prediction.
`_Have_lzcnt` is the same on each call, so never mispredicted.

Reply:
I mean, due to branch prediction it is better to call `__lzcnt` for zero if it works, this way we don't have mispredicted branches on the fast path.

Reply:
as I've said before Alex is totally correct. (well it can be misperdicted on the _first_ call or two with a cold btb)